### PR TITLE
improve imagick, bcmath and gmp extension warnings

### LINKED ISSUE #22849: Remove bcmath and gmp warning if WebAuthn is not used
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

**Is your feature request related to a problem? Please describe.**
The admin panel prints a warning if `bcmath` and/or `gmp` PHP module is not available. Those two modules are (AFAIK) only required for WebAuthn. The generic "urgent" module recommendation text is misleading in this case since it does not imply any performance or compatibility issue to not have them installed, aside of an optional non-security relevant feature being not available.

**Describe the solution you'd like**
If I followed the `isWebAuthnAvailable` function correctly, the WebAuthn section is only shown with this two modules installed. I would instead like to have the WebAuthn section always shown and the "missing modules" info shown there in case. So users and admins always see this Nextcloud feature but as well see the info if and why it cannot be used. This is especially useful for shared hostings, so users can inform the hosting provider and ask to have those modules installed to make a certain feature available, instead of saying "the admin panel recommends for some generic reason".

**Describe alternatives you've considered**
Alternatively the admin panel could print more detailed info what this modules are required for. So admins don't need to either track the reason themselves or follow the recommendation "blindly" if even possible or ignore it without having a good feeling with this (🙂) but know exactly what those modules are for and decide well-founded.

**Additional context**
Relevant code:
- Admin panel checks: https://github.com/nextcloud/server/blob/987f62117351a658dc8c0916e38cfb0b7ca60982/apps/settings/lib/Controller/CheckSetupController.php#L592-L598
- WebAuthn check: https://github.com/nextcloud/server/blob/9271d65641493454af4d79e21a500ff1da1387fc/lib/private/Authentication/WebAuthn/Manager.php#L260-L274
- Settings section: https://github.com/nextcloud/server/blob/b219eadb8f3f459f2c71e937dfd3542f0792da8e/apps/settings/lib/Settings/Personal/Security/WebAuthn.php#L71-L77

### PR DESCRIPTION
**Pull request by szaimen**
Fix part of https://github.com/nextcloud/server/issues/13099
Fix https://github.com/nextcloud/server/issues/22849

Signed-off-by: szaimen <szaimen@e.mail.de>

### CODE REVIEW

**Comment by MichaIng** [comment]
Not a real solution, but most appliance and NC installer developers will like it, including myself, muting one of the warnings we do not want to address intentionally.

**Review by MichaIng — APPROVED** [review]

**Comment by skjnldsv** [comment]
I'm really not ok with that. Yet again a random hidden config just to tell the admin everything is fine?
The admin should know how to run a nc instance and learn what this warning means and if they can ignore it or not. Adding a config switch for a cosmetic reason is not the way to go imho

**Comment by MichaIng** [comment]
As much as I would like to mute this for our users, I have to agree with @skjnldsv, but the way the warning shows up currently does not serve the stated intention:

> The admin should know how to run a nc instance and learn what this warning means

Then we should rephrase the text from a generic "highly recommended for performance and compatibility" to one that DOES explain what it is really used for. AFAIK it speeds up thumbnail, avatar and possibly other UI image resizing, compared to GD. As this is a one time step until new images are added, IMHO it is not worth it to bloat the PHP instance, considering the open security questions discussed in #13099.

Similar with `bcmath` and `gmp`, required for WebAuthn authentication but otherwise unnecessary, so the warnings are currently misleading, IMHO: #22849

**Comment by szaimen** [comment]
> Then we should rephrase the text from a generic "highly recommended for performance and compatibility" to one that DOES explain what it is really used for. AFAIK it speeds up thumbnail, avatar and possibly other UI image resizing, compared to GD. As this is a one time step until new images are added, IMHO it is not worth it to bloat the PHP instance, considering the open security questions discussed in #13099.
> 
> Similar with `bcmath` and `gmp`, required for WebAuthn authentication but otherwise unnecessary, so the warnings are currently misleading, IMHO: #22849

Sounds good to me. Any pointers how and where to implement this?

**Comment by MichaIng** [comment]
At best a dedicated function like `hasImagickPHPModule()`, passing a dedicated `DataResponse` array entry (bottom of this script), which triggers a dedicated warning in the frontend: https://github.com/nextcloud/server/blob/master/core/js/setupchecks.js

There are a few tests which would need to be adjusted as well, see the changes for OPcache setup checks: https://github.com/nextcloud/server/pull/27403/files

**Comment by szaimen** [comment]
@MichaIng does https://github.com/nextcloud/server/pull/31470/commits/70bd0083602503b7ecdb44987b0a3a9e61d9c263 look good to you? I will fix the tests afterwards.

File: core/js/setupchecks.js [hunk 1]
```diff
@@ -211,7 +211,7 @@
 					}
 					if(!data.hasFileinfoInstalled) {
 						messages.push({
-							msg: t('core', 'The PHP module "fileinfo" is missing. It is strongly recommended to enable this module to get the best results with MIME type detection.'),
+							msg: t('core', 'The PHP "fileinfo" is missing. It is strongly recommended to enable this module to get the best results with MIME type detection.'),
```
[the remark this answers is not in the document]
Comment (szaimen) [reply core/js/setupchecks.js:214]: yes

File: core/js/setupchecks.js [hunk 2]
```diff
@@ -423,6 +423,24 @@
 							type: OC.SetupChecks.MESSAGE_TYPE_INFO
 						})
 					}
+					if (!data.isImagickEnabled) {
+						messages.push({
+							msg: t(
+								'core',
+								'Imagick is not enabled although the theming app is. For Favicon generation to work correctly, you need to install the imagick PHP extension.'
```
Comment (MichaIng) [inline core/js/setupchecks.js:430]:
```suggestion
								'The PHP module "imagick" is not enabled although the theming app is. For favicon generation to work correctly, you need to install and enable this module.'
```
Just for aligning with other messages 🙂.

File: core/js/setupchecks.js [hunk 3]
```diff
@@ -423,6 +423,24 @@
...
+							),
+							type: OC.SetupChecks.MESSAGE_TYPE_INFO
+						})
+					}
+					if (!data.areWebauthnExtensionsEnabled) {
+						messages.push({
+							msg: t(
+								'core',
+								'The PHP extensions gmp or bcmath are not enabled. For Webauthn to work correctly, please enable these extensions.'
```
Comment (MichaIng) [inline core/js/setupchecks.js:439]:
```suggestion
								'The PHP modules "gmp" and/or "bcmath" are not enabled. If you use WebAuthn passwordless authentication, these modules are required.'
```
Alignment and to clarify that these are required only when WebAuthn is actually used.

**Comment by MichaIng** [comment]
LGTM, aside of some comments I left for message consistency and to make clearer that the two WebAuthn modules are required only when WebAuthn is actually used. IMHO having dedicated messages or each of these optional modules is much better than the previous generic message 👍. That way admins have a better basis to decide whether to ignore the warning or install the modules.

File: core/js/setupchecks.js (hunk 2 above)
Comment (szaimen) [reply core/js/setupchecks.js:430]: lgtm :)

File: core/js/setupchecks.js (hunk 3 above)
Comment (szaimen) [reply core/js/setupchecks.js:439]: lgtm :)

**Review by MichaIng — APPROVED** [review]
Approving, given the tests will be fixed 👍.

**Comment by szaimen** [comment]
Tests are fixed :)

**Review by skjnldsv — APPROVED** [review]
Nice!

**Comment by MichaIng** [comment]
/backport to stable23

Probably? 🙂

**Comment by szaimen** [comment]
> 

Not sure. @skjnldsv?

**Pull request merged**